qa/tasks/cephfs/test_exports.py

```diff
@@ -627,3 +627,372 @@ def test_ephemeral_pin_shrink_mds(self):
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, "2")
+        self._wait_subtrees([('/1', 2)], status=self.status)
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "-1")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, None)
+
+        self.mount_a.setfattr(".", "ceph.dir.bal.mask", "0")
+        self._wait_subtrees([('', 0)], status=self.status, path='', exclude_path='~mds')
+
+    def test_unset_rank_mask_under_nested_parent(self):
+        """
+        That ceph.dir.bal.mask is unset under nested directory.
+        """
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "1")
+        self._wait_subtrees([('/1', 1)], status=self.status, rank=1)
+        self.mount_a.setfattr("1/a", "ceph.dir.bal.mask", "2")
+        self._wait_subtrees([('/1', 1), ('/1/a', 2)], status=self.status, rank=2)
+
+        self.mount_a.setfattr("1/a", "ceph.dir.bal.mask", "-1")
+        self._wait_subtrees([('/1', 1)], status=self.status, rank=1)
+
+    def test_mds_failover(self):
+        """
+        That MDS failover does not affect the ceph.dir.bal.mask.
+        """
+        self.config_set('mds', 'mds_debug_subtrees', False)
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "2")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, "2")
+        self._wait_subtrees([('/1', 2)], status=self.status)
+
+        for i in range(5):
+            self.fs.rank_fail(rank=2)
+            self.status = self.fs.wait_for_daemons()
+
+            value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+            self.assertEqual(value, "2")
+            self._wait_subtrees([('/1', 2)], status=self.status)
+
+    def test_mds_shrink(self):
+        """
+        That ceph.dir.bal.mask is sustained during reducing MDS.
+        """
+
+        self.fs.set_max_mds(3)
+        self.status = self.fs.wait_for_daemons()
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "1")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, "1")
+        self._wait_subtrees([('/1', 1)], status=self.status)
+
+        self.fs.set_max_mds(2)
+        self.status = self.fs.wait_for_daemons()
+        self._wait_subtrees([('/1', 1)], status=self.status)
+
+    def test_mds_shrink_and_grow(self):
+        """
+        That ceph.dir.bal.mask is sustained during reducing and growing MDS.
+        """
+
+        self.fs.set_max_mds(3)
+        self.status = self.fs.wait_for_daemons()
+
+        self.mount_a.setfattr("1", "ceph.dir.bal.mask", "2")
+        value = self.mount_a.getfattr("1", "ceph.dir.bal.mask")
+        self.assertEqual(value, "2")
+        self._wait_subtrees([('/1', 2)], status=self.status)
+
+        self.fs.set_max_mds(2)
+        self.status = self.fs.wait_for_daemons()
+        self._wait_subtrees([('/1', 0)], status=self.status)
```

Comment:
why can it not be on rank 1? Is there a documented rule for this scenario?

Reply:
I understand that when max_mds shrink occurs, the subtrees in mds move to rank 0. 
https://github.com/ceph/ceph/blob/main/src/mds/MDCache.cc#L7909

Well, if there is a point that I don't understand, please leave a comment.

Reply:
I see. Please make sure the documentation addresses this.

Reply:
Added.

https://github.com/ceph/ceph/compare/9b04cd83fb615464888de665b9933d37724d4a28..ac1ab14f09a789361f9a5c72e448ae85f5b55fa8#diff-cc9cf0d49df0c5644ecb9c98d1f7ac1d307b05837bfdabc6761c65691c6f8dc1R438